### PR TITLE
task/FP-1984: Paginator component rework

### PR DESCRIPTION
**Pull request by edmondsgarrett**
## Overview
Made paginator in apcd a common component, altered its implementation and styling as well
…

## Related

- [FP-1984](https://jira.tacc.utexas.edu/browse/FP-1984)
- [FP-1973](https://jira.tacc.utexas.edu/browse/FP-1973)
…

## Changes
- Made new `components` folder for common components to live in
- Added paginator code: backend, HTML, and styling (from Core-Styles)
- Added new paginator to admin list registrations page
- Changed paginator on admin list submissions form to new component
…

## Testing

<b> List Registration Requests <b>
1. Replace lines 67-69 in `apps/admin_regis_table/views.py` with the following:
   <details><summary>Snippet</summary>
     
          import datetime
          registrations_content = [
            (
                1,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                False,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'received',                           #registration_status
                'insurance carrier',                #org_type
                'Golden Rule Insurance Company',    #business_name
                '7440 Woodland Drive',              #mail_address
                'Indianpolis',                      #city
                'IN',                               #state
                '46278     '                        #zip
            ),
            (
                2,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                False,                               #submitting_for_self
                'SFTP',                             #submission_method
                'processing',                           #registration_status
                'pbm',                #org_type
                'Clay Tenant Insurance Company',    #business_name
                '440 Wood and Drive',              #mail_address
                'Indianapolis',                      #city
                'NY',                               #state
                '24444     '                        #zip
            ),
            (
                3,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'complete',                        #registration_status
                'insurance carrier',                #org_type
                'Magenta Condition Insurance Company',    #business_name
                '742 Oddland Drive',              #mail_address
                'Indipolis',                      #city
                'MI',                               #state
                '62784     '                        #zip
            ),
            (
                1,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                False,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'received',                           #registration_status
                'insurance carrier',                #org_type
                'Golden Rule Insurance Company',    #business_name
                '7440 Woodland Drive',              #mail_address
                'Indianpolis',                      #city
                'IN',                               #state
                '46278     '                        #zip
            ),
            (
                2,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                False,                               #submitting_for_self
                'SFTP',                             #submission_method
                'processing',                           #registration_status
                'pbm',                #org_type
                'Clay Tenant Insurance Company',    #business_name
                '440 Wood and Drive',              #mail_address
                'Indianapolis',                      #city
                'NY',                               #state
                '24444     '                        #zip
            ),
            (
                3,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'complete',                        #registration_status
                'insurance carrier',                #org_type
                'Magenta Condition Insurance Company',    #business_name
                '742 Oddland Drive',              #mail_address
                'Indipolis',                      #city
                'MI',                               #state
                '62784     '                        #zip
            ),
            (
                1,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                False,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'received',                           #registration_status
                'insurance carrier',                #org_type
                'Golden Rule Insurance Company',    #business_name
                '7440 Woodland Drive',              #mail_address
                'Indianpolis',                      #city
                'IN',                               #state
                '46278     '                        #zip
            ),
            (
                2,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                False,                               #submitting_for_self
                'SFTP',                             #submission_method
                'processing',                           #registration_status
                'pbm',                #org_type
                'Clay Tenant Insurance Company',    #business_name
                '440 Wood and Drive',              #mail_address
                'Indianapolis',                      #city
                'NY',                               #state
                '24444     '                        #zip
            ),
            (
                3,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'complete',                        #registration_status
                'insurance carrier',                #org_type
                'Magenta Condition Insurance Company',    #business_name
                '742 Oddland Drive',              #mail_address
                'Indipolis',                      #city
                'MI',                               #state
                '62784     '                        #zip
            ),
            (
                1,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                False,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'received',                           #registration_status
                'insurance carrier',                #org_type
                'Golden Rule Insurance Company',    #business_name
                '7440 Woodland Drive',              #mail_address
                'Indianpolis',                      #city
                'IN',                               #state
                '46278     '                        #zip
            ),
            (
                2,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                False,                               #submitting_for_self
                'SFTP',                             #submission_method
                'processing',                           #registration_status
                'pbm',                #org_type
                'Clay Tenant Insurance Company',    #business_name
                '440 Wood and Drive',              #mail_address
                'Indianapolis',                      #city
                'NY',                               #state
                '24444     '                        #zip
            ),
            (
                3,                                  #registration_id
                datetime.date(2022, 8, 3),          #posted_date
                12023,                              #applicable_period_start
                122023,                             #applicable_period_end
                True,                               #file_me
                True,                               #file_pv
                True,                               #file_mc
                True,                               #file_pc
                False,                              #file_dc
                True,                               #submitting_for_self
                'SFTP',                             #submission_method
                'complete',                        #registration_status
                'insurance carrier',                #org_type
                'Magenta Condition Insurance Company',    #business_name
                '742 Oddland Drive',              #mail_address
                'Indipolis',                      #city
                'MI',                               #state
                '62784     '                        #zip
            )
        ]
        registrations_entities = [
            (
                5, 1, 5, 46, 11111, None, 5, 'Garretts Test Business 2', '11-0001111'
            ),
            (
                5, 2, 50000, 47, 11010, 21101, 1, 'A Second Test Entity', '00-0000000'
            )
        ]
        registrations_contacts = [
            (
                52,
                1,
                False,
                'Test Role',
                'Garrett Test Tester',
                '2222222222',
                'notarealemail@emailplace.com'
            ),
            (
                53,
                1,
                False,
                'A 2nd Test Role',
                'Test Garrett 2 Test',
                '15555555555',
                'testemail@testemail.net'
            ),
            (
                54,
                2,
                True,
                'A 3rd and final test role',
                'Test 3rd Role Name Garrett',
                '0000000000',
                'email@gmail.com'
            )
          ]
        
    </details>
2. Change line 236 in the same file to `context.update(paginator(self.request, registration_table_entries, 1))`
3. Open http://localhost:8000/administration/list-registration-requests/
4. Ensure paginator appears under the table as shown in screenshot below, and that clicking through pages show different data, correctly updates URL with page number, and updates paginator

<b> List Submissions - Admin <b>
1. In `apcd_cms/src/apps/admin_submissions/views.py`, make the following changes:

<details>
       <summary>View Snippet</summary>

Add the following to the top of the file:

`import datetime`

 Replace line 26 with the following:

  ```js
submission_content = [
      (
          1,                                                      # submission_id
          1,                                                      # apcd_id
          1,                                                      # submitter_id
          'T_GLDNRLE_10000000_202303_202303.zip',                 # zip_file_name
          'complete',                                             # status
          'failed',                                               # outcome
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # received_timestamp
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # updated_at
          'CHCD'                                                  # org_name
      ),
      (
          1,                                                      # submission_id
          1,                                                      # apcd_id
          1,                                                      # submitter_id
          'T_GLDNRLE_10000000_202303_202303.zip',                 # zip_file_name
          'complete',                                             # status
          'failed',                                               # outcome
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # received_timestamp
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # updated_at
          'CHCD'                                                  # org_name
      ),
      (
          1,                                                      # submission_id
          1,                                                      # apcd_id
          1,                                                      # submitter_id
          'T_GLDNRLE_10000000_202303_202303.zip',                 # zip_file_name
          'complete',                                             # status
          'failed',                                               # outcome
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # received_timestamp
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # updated_at
          'CHCD'                                                  # org_name
      ),
      (
          1,                                                      # submission_id
          1,                                                      # apcd_id
          1,                                                      # submitter_id
          'T_GLDNRLE_10000000_202303_202303.zip',                 # zip_file_name
          'complete',                                             # status
          'failed',                                               # outcome
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # received_timestamp
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # updated_at
          'CHCD'                                                  # org_name
      ),
      (
          1,                                                      # submission_id
          1,                                                      # apcd_id
          1,                                                      # submitter_id
          'T_GLDNRLE_10000000_202303_202303.zip',                 # zip_file_name
          'complete',                                             # status
          'failed',                                               # outcome
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # received_timestamp
          datetime.datetime(2022, 9, 27, 15, 25, 24, 285442),     # updated_at
          'CHCD'                                                  # org_name
      ),
  ]
  ```

 Replace line 63 with the following:

  ```js
submission_logs = [
      (
          1,                          # log_id
          1,                          # submission_id
          '  ',                       # file_type
          'init_txt_file_suite',      # validation_suite
          {'test': 'here'},           # json_long
          'failed',                   # outcome
          'Medical Provider'
      ),
      (
          2,                          # log_id
          1,                          # submission_id
          'ME',                       # file_type
          'eligibility_suite',        # validation_suite
          {'test': 'here'},           # json_long
          'failed',                   # outcome
          'Provider'  
      ),
      (
          3,                          # log_id
          1,                          # submission_id
          'PC',                       # file_type
          'pharmacy_suite',           # validation_suite
          {'test': 'here'},           # json_long
          'failed',                   # outcome
          'Pharmacy Claims'  
      ),
      (
          4,                          # log_id
          1,                          # submission_id
          'PV',                       # file_type
          'provider_suite',           # validation_suite
          {'test': 'here'},           # json_long
          'failed',                   # outcome
          'Medical Provider'  
      )
  ]
  ```
  Replace line 66 with: `context.update(paginator(self.request, submission_with_logs, 1))`
</details>

2. Open [http://localhost:8000/administration/list-submissions](http://localhost:8000/administration/list-submissions)
3. Test paginator as you did for the previous page

## UI
image

…

### CODE REVIEW

**Review by jmartinez-tacc — APPROVED** [review]
LGTM!

File: apcd-cms/src/apps/admin_submissions/views.py [hunk 1]
```diff
@@ -20,6 +21,7 @@ def dispatch(self, request, *args, **kwargs):
     def get_context_data(self, *args, **kwargs):
 
         context = super(AdminSubmissionsTable, self).get_context_data(*args, **kwargs)
+        import datetime
```
Comment (shayanaijaz) [inline apcd-cms/src/apps/admin_submissions/views.py:24]: leftover from testing?

**Review by shayanaijaz — COMMENTED** [review]
Looks really good and works consistently! Only thought I had was maybe we should keep the number of pages appearing in the pagination the same throughout? For example, having a constant 5 pages being shown (given there are 5 or more than 5 pages). That way we would not have the pagination constantly changing sizes. Let me know what you think!

File: apcd-cms/src/apps/admin_submissions/views.py (hunk 1 above)
Comment (edmondsgarrett) [reply apcd-cms/src/apps/admin_submissions/views.py:24]: Will remove, thanks for catching that

**Comment by edmondsgarrett** [comment]
> Looks really good and works consistently! Only thought I had was maybe we should keep the number of pages appearing in the pagination the same throughout? For example, having a constant 5 pages being shown (given there are 5 or more than 5 pages). That way we would not have the pagination constantly changing sizes. Let me know what you think!

I could see the benefit of that! I am curious what Wes thinks about this.

File: apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+.pagination {
+    margin-block:var(--global-space--x-large);
+    margin-inline:auto;
+    width:-moz-fit-content;
+    width:fit-content
+}
+/*.pagination a,.pagination span {
+    padding-block:0
+}*/
```
Comment (wesleyboar) [inline apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css:70]: **Undesired Opportunity for Untracked Divergence from Source.** By pasting in an _un-minified_ stylesheet, you've suggested to other developers that these styles are APCD-specific. Which then leads to edits to the styles that are not tracked as differences with the source.

Please treat these styles like external third-party code[^1], and customize them using (A) the available customization options and/or (B) overwriting the styles.

<details><summary>A. <strong>Available Customization</strong></summary>

I.e. [CSS Custom Properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*). Not suitable for the removal of `padding-block`, so see **B**.
```css
    --item-width
    --item-buffer
    --item-border
```

</details>
<details><summary>B. <strong>Overwriting Styles</strong></summary>

There are multiple ways to do this clearly. Evaluate the URL to use in these examples, after reading [my other comment about these styles].

1. ```css
    /*! @tacc/core-styles 0.10.0+ | MIT | github.com/TACC/Core-Styles */.c-page-list,.pagination{--item-width:2em;--item-buffer:3px;--item-border:var(--global-border-width--normal)}.c-page-list,.pagination{align-items:baseline;display:flex;flex-direction:row;font-size:14px;list-style:none}ul.c-page-list{list-style:none}.c-page-link--always-click{box-sizing:content-box;position:relative}.c-page-link--always-click:after,.c-page-link--always-click:before{--width:var(--item-buffer);--vert-offset:calc(var(--item-border)*-1);--horz-offset:calc(100% + var(--item-border));bottom:var(--vert-offset);content:"";position:absolute;top:var(--vert-offset);width:var(--width)}.c-page-link--always-click.c-button{overflow:visible}.c-page-link--always-click:before{right:var(--horz-offset)}.c-page-link--always-click:after{left:var(--horz-offset)}.c-page-item{margin-inline:var(--item-buffer)}.c-page-item:first-child{margin-left:0}.c-page-item:last-child{margin-right:0}.c-page-item--etcetera{cursor:default;margin-inline:2px;min-width:var(--item-width);text-align:center}.c-page-end,.pagination a,.pagination span{padding:4px 12px}.c-page-link{line-height:1.2;min-width:var(--item-width);padding:5px 0}.pagination{margin-block:var(--global-space--x-large);margin-inline:auto;width:-moz-fit-content;width:fit-content}.pagination a,.pagination span{padding-block:0}

    .pagination a,
    .pagination span {
        padding-block: unset; /* or some other suitable value */
    }
    ```
2. ```css
    @import url("https://unpkg.com/browse/@tacc/core-styles@0.10.0/dist/components/bootstrap.pagination.css");

    .pagination a,
    .pagination span {
        padding-block: unset; /* or some suitable value */
    }
    ```

To learn more about `unset` and friends, see https://css-tricks.com/inherit-initial-unset-revert/.

</details>

[^1]: These are from an outside library: https://www.npmjs.com/package/@tacc/core-styles.

Comment (wesleyboar) [inline apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css:70]: **Wes Confused.** Why is this code commented out? I do not see `pagination` class even used in the markup. Did I miss it?

File: apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/*! @tacc/core-styles 1.0.0+ | MIT | github.com/TACC/Core-Styles */
```
Comment (wesleyboar) [inline apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css:1]: **Superfluous Style Load?** The file `apcd-cms/src/apps/components/paginator/templates/paginator.html` uses `c-page` classes. **If** the Bootstrap `pagaination` class are not used at all, **then** please only load the Core-Styles for the markup you use:

https://unpkg.com/@tacc/core-styles@0.10.0/dist/components/c-page.css

**Review by wesleyboar — COMMENTED** [review]
I have concerns with how CSS from an external library is used.

_In these concerns I may reference Core-Styles v0.10.0 instead of Core-Styles v1.0.0, because v0.10 is the version that the APCD's Core-CMS image uses.[^1]_

File: apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css (hunk 2 above)
Comment (wesleyboar) [reply apcd-cms/src/apps/components/paginator/static/paginator/css/paginator.css:70]: It was my original idea to import an un-minified stylesheet, so I caused this "undesired" thing 😅.

At the time, I was unaware that unpkg.com had all npm package version files available.

**Review by shayanaijaz — APPROVED** [review]
Apart from any other potential css changes, it looks good to me

**Comment by wesleyboar** [comment]
@shayanaijaz Garret mentioned your idea to me. I like that you're thinking about this. There is precedent for a similar idea.

The https://github.com/TACC/Core-Portal's Pagination component only shows 11 page max, but does show a variable number of pages per usage before reaching that many pages ([spec](https://confluence.tacc.utexas.edu/display/UP/UI+-+Components+-+Pagination)).

Until pages number up to 11, I think we need not worry about a constant number of a maximum. If the user experience on APCD is different in a way that you propose a different strategy, please propose or inquire in our Slack channel `#cep-design`.

**Review by wesleyboar — COMMENTED** [review]
✅ My PR approval is because of 2b9c9e2.

**Pull request merged**